File: app/manifest_service.py
from __future__ import annotations

import re
from pathlib import Path

from app.models import ApkInfo, ModPayload, Task
from app.task_store import log_task
from app.validators import is_valid_package_name, is_valid_version_code


def _escape_xml_attr(value: str) -> str:
    return (
        value.replace('&', '&amp;')
        .replace('"', '&quot;')
        .replace('<', '&lt;')
        .replace('>', '&gt;')
    )


def _read_attr(tag: str, attr: str) -> str:
    escaped = re.escape(attr)
    m = re.search(rf'{escaped}="([^"]*)"', tag)
    return m.group(1) if m else ''


def _decode_resource_string(value: str) -> str:
    return (
        value.replace('&quot;', '"')
        .replace('&apos;', "'")
        .replace('&lt;', '<')
        .replace('&gt;', '>')
        .replace('&amp;', '&')
    )


def _read_version_from_apktool_yml(decoded_dir: Path) -> tuple[str, str]:
    yml_path = decoded_dir / 'apktool.yml'
    if not yml_path.exists():
        return '', ''
    text = yml_path.read_text(encoding='utf-8')

    def _extract(name: str) -> str:
        m1 = re.search(rf'^\s*{name}:\s*[\'\"]?([^\r\n\'\"]+)[\'\"]?\s*$', text, flags=re.MULTILINE)
        if m1:
            return m1.group(1).strip()
        m2 = re.search(rf'^\s*{name}:\s*([^\r\n]+)\s*$', text, flags=re.MULTILINE)
        return m2.group(1).replace("'", '').replace('"', '').strip() if m2 else ''

    return _extract('versionName'), _extract('versionCode')


def _resolve_string_label(decoded_dir: Path, label_ref: str) -> str:
    if not label_ref.startswith('@string/'):
        return label_ref

    name = label_ref.replace('@string/', '').strip()
    if not name:
        return label_ref

    res_dir = decoded_dir / 'res'
    dirs = ['values'] + [x.name for x in res_dir.iterdir()] if res_dir.exists() else ['values']
    values_dirs = []
    for item in dirs:
        if item.startswith('values') and item not in values_dirs:
            values_dirs.append(item)

    escaped_name = re.escape(name)
    for values_dir in values_dirs:
        strings_path = decoded_dir / 'res' / values_dir / 'strings.xml'
        if not strings_path.exists():
            continue
        content = strings_path.read_text(encoding='utf-8')
        m = re.search(rf'<string\s+name="{escaped_name}"[^>]*>([\s\S]*?)</string>', content)
        if m and m.group(1):
            return _decode_resource_string(m.group(1).strip())

    return label_ref


def _resolve_icon_file(decoded_dir: Path, icon_ref: str) -> str | None:
    if not icon_ref.startswith('@'):
        return None

    clean = icon_ref[1:]
    if '/' not in clean:
        return None
    res_type, res_name = clean.split('/', 1)
    if not res_type or not res_name:
        return None

    res_root = decoded_dir / 'res'
    if not res_root.exists():
        return None

    density_rank = ['xxxhdpi', 'xxhdpi', 'xhdpi', 'hdpi', 'mdpi', 'anydpi']
    candidates: list[Path] = []
    for folder in res_root.iterdir():
        if not folder.is_dir():
            continue
        if folder.name != res_type and not folder.name.startswith(f'{res_type}-'):
            continue
        for child in folder.iterdir():
            if child.stem == res_name and child.suffix.lower() in {'.png', '.webp', '.jpg', '.jpeg'}:
                candidates.append(child)

    if not candidates:
        return None

    def _score(p: Path) -> int:
        directory = p.parent.name
        for idx, key in enumerate(density_rank):
            if key in directory:
                return idx
        return len(density_rank)

    candidates.sort(key=_score)
    return str(candidates[0])


def parse_apk_info(task: Task) -> None:
    if not task.decodedDir:
        return
    decoded_dir = Path(task.decodedDir)
    manifest_path = decoded_dir / 'AndroidManifest.xml'
    if not manifest_path.exists():
        return

    xml = manifest_path.read_text(encoding='utf-8')
    manifest_tag = re.search(r'<manifest\b[^>]*>', xml)
    app_tag = re.search(r'<application\b[^>]*>', xml)

    manifest_text = manifest_tag.group(0) if manifest_tag else ''
    app_text = app_tag.group(0) if app_tag else ''

    package_name = _read_attr(manifest_text, 'package')
    version_name = _read_attr(manifest_text, 'android:versionName')
    version_code = _read_attr(manifest_text, 'android:versionCode')
    app_label_raw = _read_attr(app_text, 'android:label')
    icon_ref = _read_attr(app_text, 'android:icon')

    app_name = _resolve_string_label(decoded_dir, app_label_raw or package_name or '')
    icon_file_path = _resolve_icon_file(decoded_dir, icon_ref)
    if not version_name or not version_code:
        vn, vc = _read_version_from_apktool_yml(decoded_dir)
        version_name = version_name or vn
        version_code = version_code or vc

    task.iconFilePath = icon_file_path
    task.apkInfo = ApkInfo(
        appName=app_name,
        packageName=package_name,
        versionName=version_name,
        versionCode=version_code,
        appLabelRaw=app_label_raw,
        iconRef=icon_ref,
        iconUrl=f'/api/icon/{task.id}?v=1' if icon_file_path else None,
    )


def apply_icon_replacement(task: Task, icon_upload_path: str, ext: str) -> str:
    if not task.decodedDir:
        raise RuntimeError('Decoded directory is missing')

    res_root = Path(task.decodedDir) / 'res'
    res_root.mkdir(parents=True, exist_ok=True)

    target_name = 'ic_launcher_modder'
    mipmap_dirs = [d.name for d in res_root.iterdir() if d.is_dir() and d.name.startswith('mipmap')] if res_root.exists() else []
    if not mipmap_dirs:
        mipmap_dirs = ['mipmap-mdpi', 'mipmap-hdpi', 'mipmap-xhdpi', 'mipmap-xxhdpi', 'mipmap-xxxhdpi']

    source = Path(icon_upload_path)
    for folder in mipmap_dirs:
        folder_path = res_root / folder
        folder_path.mkdir(parents=True, exist_ok=True)
        (folder_path / f'{target_name}{ext}').write_bytes(source.read_bytes())

    log_task(task, f'Icon files updated in {len(mipmap_dirs)} mipmap folders')
    return f'@mipmap/{target_name}'


def update_manifest(task: Task, payload: ModPayload, icon_ref: str | None = None) -> None:
    if not task.decodedDir:
        raise RuntimeError('Decoded directory is missing')

    manifest_path = Path(task.decodedDir) / 'AndroidManifest.xml'
    if not manifest_path.exists():
        raise RuntimeError('AndroidManifest.xml not found after decompile')

    xml = manifest_path.read_text(encoding='utf-8')
    current_manifest_tag = re.search(r'<manifest\b[^>]*>', xml)
    current_package_name = _read_attr(current_manifest_tag.group(0) if current_manifest_tag else '', 'package')

    if payload.packageName:
        package_name = payload.packageName.strip()
        if not is_valid_package_name(package_name):
            raise RuntimeError('Invalid package name format')

        if re.search(r'\bpackage="[^"]*"', xml):
            xml = re.sub(r'\bpackage="[^"]*"', f'package="{_escape_xml_attr(package_name)}"', xml, count=1)
        elif re.search(r'<manifest\b[^>]*>', xml):
            xml = re.sub(
                r'<manifest\b([^>]*)>',
                f'<manifest\\1 package="{_escape_xml_attr(package_name)}">',
                xml,
                count=1,
            )
        log_task(task, f'Manifest updated: packageName={package_name}')

        if current_package_name and current_package_name != package_name:
            old_prefix = re.escape(f'{current_package_name}.')
            xml = re.sub(old_prefix, f'{package_name}.', xml)
            xml = re.sub(
                rf'([="\']){re.escape(current_package_name)}(["\'])',
                rf'\1{package_name}\2',
                xml,
            )
            log_task(task, f'Manifest updated: renamed package-scoped refs {current_package_name} -> {package_name}')

    if payload.versionName:
        version_name = _escape_xml_attr(payload.versionName.strip())
        if re.search(r'android:versionName="[^"]*"', xml):
            xml = re.sub(r'android:versionName="[^"]*"', f'android:versionName="{version_name}"', xml, count=1)
        elif re.search(r'<manifest\b[^>]*>', xml):
            xml = re.sub(r'<manifest\b([^>]*)>', f'<manifest\\1 android:versionName="{version_name}">', xml, count=1)
        log_task(task, f'Manifest updated: versionName={payload.versionName}')

    if payload.versionCode:
        version_code = payload.versionCode.strip()
        if not is_valid_version_code(version_code):
            raise RuntimeError('versionCode must be a non-negative integer')
        if re.search(r'android:versionCode="[^"]*"', xml):
            xml = re.sub(r'android:versionCode="[^"]*"', f'android:versionCode="{version_code}"', xml, count=1)
        elif re.search(r'<manifest\b[^>]*>', xml):
            xml = re.sub(r'<manifest\b([^>]*)>', f'<manifest\\1 android:versionCode="{version_code}">', xml, count=1)
        log_task(task, f'Manifest updated: versionCode={payload.versionCode}')

    if payload.appName:
        app_name = _escape_xml_attr(payload.appName.strip())
        if re.search(r'android:label="[^"]*"', xml):
            xml = re.sub(r'android:label="[^"]*"', f'android:label="{app_name}"', xml, count=1)
        elif re.search(r'<application\b[^>]*>', xml):
            xml = re.sub(r'<application\b([^>]*)>', f'<application\\1 android:label="{app_name}">', xml, count=1)
        else:
            raise RuntimeError('No <application> tag found in AndroidManifest.xml')
        log_task(task, f'Manifest updated: appName={payload.appName}')

    if icon_ref:
        safe_icon = _escape_xml_attr(icon_ref)
        if re.search(r'android:icon="[^"]*"', xml):
            xml = re.sub(r'android:icon="[^"]*"', f'android:icon="{safe_icon}"', xml, count=1)
        elif re.search(r'<application\b[^>]*>', xml):
            xml = re.sub(r'<application\b([^>]*)>', f'<application\\1 android:icon="{safe_icon}">', xml, count=1)

        if re.search(r'android:roundIcon="[^"]*"', xml):
            xml = re.sub(r'android:roundIcon="[^"]*"', f'android:roundIcon="{safe_icon}"', xml, count=1)
        log_task(task, f'Manifest updated: iconRef={icon_ref}')

    manifest_path.write_text(xml, encoding='utf-8')

File: src/manifestService.ts
import fs from 'fs';
import path from 'path';
import { logTask } from './taskStore';
import { ModPayload, Task } from './types';
import { isValidPackageName, isValidVersionCode } from './validators';

function escapeXmlAttr(value: string): string {
  return value.replace(/&/g, '&amp;').replace(/"/g, '&quot;').replace(/</g, '&lt;').replace(/>/g, '&gt;');
}

function readAttr(tag: string, attr: string): string {
  const escaped = attr.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  const match = tag.match(new RegExp(`${escaped}="([^"]*)"`));
  return match?.[1] || '';
}

function decodeResourceString(value: string): string {
  return value
    .replace(/&quot;/g, '"')
    .replace(/&apos;/g, "'")
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&amp;/g, '&');
}

function readVersionFromApktoolYml(decodedDir: string): { versionName: string; versionCode: string } {
  const ymlPath = path.join(decodedDir, 'apktool.yml');
  if (!fs.existsSync(ymlPath)) {
    return { versionName: '', versionCode: '' };
  }
  const text = fs.readFileSync(ymlPath, 'utf8');
  const extract = (name: string): string => {
    const first = text.match(new RegExp(`^\\s*${name}:\\s*['"]?([^\\r\\n'"]+)['"]?\\s*$`, 'm'));
    if (first?.[1]) {
      return first[1].trim();
    }
    const fallback = text.match(new RegExp(`^\\s*${name}:\\s*([^\\r\\n]+)\\s*$`, 'm'));
    return fallback?.[1]?.replace(/['"]/g, '').trim() || '';
  };
  return { versionName: extract('versionName'), versionCode: extract('versionCode') };
}

function resolveStringLabel(decodedDir: string, labelRef: string): string {
  if (!labelRef.startsWith('@string/')) {
    return labelRef;
  }
  const name = labelRef.replace('@string/', '').trim();
  if (!name) {
    return labelRef;
  }
  const resDir = path.join(decodedDir, 'res');
  const dirs = fs.existsSync(resDir) ? fs.readdirSync(resDir).filter(item => item.startsWith('values')) : ['values'];
  for (const valuesDir of dirs) {
    const stringsPath = path.join(resDir, valuesDir, 'strings.xml');
    if (!fs.existsSync(stringsPath)) {
      continue;
    }
    const content = fs.readFileSync(stringsPath, 'utf8');
    const escapedName = name.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
    const match = content.match(new RegExp(`<string\\s+name="${escapedName}"[^>]*>([\\s\\S]*?)</string>`));
    if (match?.[1]) {
      return decodeResourceString(match[1].trim());
    }
  }
  return labelRef;
}

function resolveIconFile(decodedDir: string, iconRef: string): string | undefined {
  if (!iconRef.startsWith('@')) {
    return undefined;
  }
  const clean = iconRef.slice(1);
  const parts = clean.split('/', 2);
  if (parts.length !== 2) {
    return undefined;
  }
  const [resType, resName] = parts;
  const resRoot = path.join(decodedDir, 'res');
  if (!fs.existsSync(resRoot)) {
    return undefined;
  }

  const densityRank = ['xxxhdpi', 'xxhdpi', 'xhdpi', 'hdpi', 'mdpi', 'anydpi'];
  const candidates: string[] = [];
  for (const folder of fs.readdirSync(resRoot)) {
    const folderPath = path.join(resRoot, folder);
    if (!fs.statSync(folderPath).isDirectory()) {
      continue;
    }
    if (folder !== resType && !folder.startsWith(`${resType}-`)) {
      continue;
    }
    for (const child of fs.readdirSync(folderPath)) {
      const ext = path.extname(child).toLowerCase();
      if (path.basename(child, ext) === resName && ['.png', '.webp', '.jpg', '.jpeg'].includes(ext)) {
        candidates.push(path.join(folderPath, child));
      }
    }
  }

  candidates.sort((left, right) => {
    const score = (target: string): number => {
      const directory = path.basename(path.dirname(target));
      const index = densityRank.findIndex(item => directory.includes(item));
      return index >= 0 ? index : densityRank.length;
    };
    return score(left) - score(right);
  });

  return candidates[0];
}

export function parseApkInfo(task: Task): Task {
  if (!task.decodedDir) {
    return task;
  }
  const manifestPath = path.join(task.decodedDir, 'AndroidManifest.xml');
  if (!fs.existsSync(manifestPath)) {
    return task;
  }

  const xml = fs.readFileSync(manifestPath, 'utf8');
  const manifestTag = xml.match(/<manifest\b[^>]*>/)?.[0] || '';
  const appTag = xml.match(/<application\b[^>]*>/)?.[0] || '';
  let versionName = readAttr(manifestTag, 'android:versionName');
  let versionCode = readAttr(manifestTag, 'android:versionCode');
  const packageName = readAttr(manifestTag, 'package');
  const appLabelRaw = readAttr(appTag, 'android:label');
  const iconRef = readAttr(appTag, 'android:icon');

  if (!versionName || !versionCode) {
    const fallback = readVersionFromApktoolYml(task.decodedDir);
    versionName ||= fallback.versionName;
    versionCode ||= fallback.versionCode;
  }

  task.iconFilePath = resolveIconFile(task.decodedDir, iconRef) || null;
  task.apkInfo = {
    appName: resolveStringLabel(task.decodedDir, appLabelRaw || packageName || ''),
    packageName,
    versionName,
    versionCode,
    appLabelRaw,
    iconRef,
    iconUrl: task.iconFilePath ? `/api/icon/${task.id}?v=1` : null,
  };
  return task;
}

export function applyIconReplacement(task: Task, iconUploadPath: string, ext: string): string {
  if (!task.decodedDir) {
    throw new Error('Decoded directory is missing');
  }
  const resRoot = path.join(task.decodedDir, 'res');
  fs.mkdirSync(resRoot, { recursive: true });
  const targetName = 'ic_launcher_modder';
  let mipmapDirs = fs
    .readdirSync(resRoot, { withFileTypes: true })
    .filter(entry => entry.isDirectory() && entry.name.startsWith('mipmap'))
    .map(entry => entry.name);
  if (mipmapDirs.length === 0) {
    mipmapDirs = ['mipmap-mdpi', 'mipmap-hdpi', 'mipmap-xhdpi', 'mipmap-xxhdpi', 'mipmap-xxxhdpi'];
  }

  const source = fs.readFileSync(iconUploadPath);
  for (const folder of mipmapDirs) {
    const folderPath = path.join(resRoot, folder);
    fs.mkdirSync(folderPath, { recursive: true });
    fs.writeFileSync(path.join(folderPath, `${targetName}${ext}`), source);
  }
  logTask(task, `Icon files updated in ${mipmapDirs.length} mipmap folders`);
  return '@mipmap/ic_launcher_modder';
}

export function updateManifest(task: Task, payload: ModPayload, iconRef?: string | null): Task {
  if (!task.decodedDir) {
    throw new Error('Decoded directory is missing');
  }
  const manifestPath = path.join(task.decodedDir, 'AndroidManifest.xml');
  if (!fs.existsSync(manifestPath)) {
    throw new Error('AndroidManifest.xml not found after decompile');
  }

  let xml = fs.readFileSync(manifestPath, 'utf8');
  const currentManifestTag = xml.match(/<manifest\b[^>]*>/)?.[0] || '';
  const currentPackageName = readAttr(currentManifestTag, 'package');

  if (payload.packageName) {
    const packageName = payload.packageName.trim();
    if (!isValidPackageName(packageName)) {
      throw new Error('Invalid package name format');
    }
    if (/\bpackage="[^"]*"/.test(xml)) {
      xml = xml.replace(/\bpackage="[^"]*"/, `package="${escapeXmlAttr(packageName)}"`);
    } else {
      xml = xml.replace(/<manifest\b([^>]*)>/, `<manifest$1 package="${escapeXmlAttr(packageName)}">`);
    }
    logTask(task, `Manifest updated: packageName=${packageName}`);
    if (currentPackageName && currentPackageName !== packageName) {
      const escaped = currentPackageName.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
      xml = xml.replace(new RegExp(`${escaped}\\.`, 'g'), `${packageName}.`);
      xml = xml.replace(new RegExp(`([="'])${escaped}(["'])`, 'g'), `$1${packageName}$2`);
      logTask(task, `Manifest updated: renamed package-scoped refs ${currentPackageName} -> ${packageName}`);
    }
  }

  if (payload.versionName) {
    const versionName = escapeXmlAttr(payload.versionName.trim());
    if (/android:versionName="[^"]*"/.test(xml)) {
      xml = xml.replace(/android:versionName="[^"]*"/, `android:versionName="${versionName}"`);
    } else {
      xml = xml.replace(/<manifest\b([^>]*)>/, `<manifest$1 android:versionName="${versionName}">`);
    }
    logTask(task, `Manifest updated: versionName=${payload.versionName}`);
  }

  if (payload.versionCode) {
    const versionCode = payload.versionCode.trim();
    if (!isValidVersionCode(versionCode)) {
      throw new Error('versionCode must be a non-negative integer');
    }
    if (/android:versionCode="[^"]*"/.test(xml)) {
      xml = xml.replace(/android:versionCode="[^"]*"/, `android:versionCode="${versionCode}"`);
    } else {
      xml = xml.replace(/<manifest\b([^>]*)>/, `<manifest$1 android:versionCode="${versionCode}">`);
    }
    logTask(task, `Manifest updated: versionCode=${payload.versionCode}`);
  }

  if (payload.appName) {
    const appName = escapeXmlAttr(payload.appName.trim());
    if (/android:label="[^"]*"/.test(xml)) {
      xml = xml.replace(/android:label="[^"]*"/, `android:label="${appName}"`);
    } else if (/<application\b[^>]*>/.test(xml)) {
      xml = xml.replace(/<application\b([^>]*)>/, `<application$1 android:label="${appName}">`);
    } else {
      throw new Error('No <application> tag found in AndroidManifest.xml');
    }
    logTask(task, `Manifest updated: appName=${payload.appName}`);
  }

  if (iconRef) {
    const safeIcon = escapeXmlAttr(iconRef);
    if (/android:icon="[^"]*"/.test(xml)) {
      xml = xml.replace(/android:icon="[^"]*"/, `android:icon="${safeIcon}"`);
    } else {
      xml = xml.replace(/<application\b([^>]*)>/, `<application$1 android:icon="${safeIcon}">`);
    }
    if (/android:roundIcon="[^"]*"/.test(xml)) {
      xml = xml.replace(/android:roundIcon="[^"]*"/, `android:roundIcon="${safeIcon}"`);
    }
    logTask(task, `Manifest updated: iconRef=${iconRef}`);
  }

  fs.writeFileSync(manifestPath, xml, 'utf8');
  return task;
}
